### PR TITLE
Implement s4u abuse

### PR DESCRIPTION
**Pull request by zblurx**
Added `--delegate` option to implement s4u abuse over smb
image
This option will do a full S4U abuse (S4U2Self + S4U2Proxy) in an automated way, allowing to use all postex functionalities of NXC :fire:

### CODE REVIEW

**Comment by bongobongoland** [comment]
omg.

**Comment by Marshall-Hallenbeck** [comment]
@zblurx can you add this to the e2e tests command file as well?

**Comment by XiaoliChan** [comment]
I test it, it works (with protocol transition)
image

**Comment by zblurx** [comment]
As @XiaoliChan suggested to me, we should also support s4u2self only:
image

**Comment by XiaoliChan** [comment]
Really awesome!
image

image

File: nxc/protocols/smb/kerberos.py
Comment (NeffIsBack) [inline nxc/protocols/smb/kerberos.py:30]: That should probably be logged through the nxc logger

File: nxc/protocols/smb/kerberos.py
Comment (NeffIsBack) [inline nxc/protocols/smb/kerberos.py:65]: Same for this, or is there a reason we would need another logger?

File: nxc/protocols/smb/kerberos.py
Comment (NeffIsBack) [inline nxc/protocols/smb/kerberos.py:261]: Is this for future use? Or old code that could be removed?

If this is for future implementation I think we could add a comment at the top clarifying it.

File: nxc/protocols/smb/kerberos.py
Comment (NeffIsBack) [inline nxc/protocols/smb/kerberos.py:354]: The indentation is one tab to much

File: nxc/protocols/smb/proto_args.py
Comment (NeffIsBack) [inline nxc/protocols/smb/proto_args.py:6]: I think for using this, `--delegation` should be required. See ldap protoargs, "no_smb_arg"

**Review by NeffIsBack — CHANGES_REQUESTED** [review]
Can you also replace the single quotes with double quotes?

Left some comments, the rest of the code looks good👍🏼

**Review by Marshall-Hallenbeck — CHANGES_REQUESTED** [review]

File: nxc/protocols/smb/kerberos.py
Comment (zblurx) [reply nxc/protocols/smb/kerberos.py]: resolved

File: nxc/protocols/smb/kerberos.py
Comment (zblurx) [reply nxc/protocols/smb/kerberos.py]: removed

File: nxc/protocols/smb/proto_args.py
Comment (zblurx) [reply nxc/protocols/smb/proto_args.py]: good idea

File: nxc/protocols/smb/kerberos.py
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:88]: Not necessary for merge, but it looks like the replace all replaced on that shouldn't be lol.

In the future the Ruff linter will properly fix these (with --fix) or let people know.

File: nxc/protocols/smb/kerberos.py
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:168]: Pythonic name should be `ticket_tgt`

File: nxc/protocols/smb/kerberos.py
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:187]: Pythonic name should be `client_name`

File: nxc/protocols/smb/kerberos.py
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:175]: Pythonic name should be `ap_req`

File: nxc/protocols/smb/kerberos.py
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:202]: Needs snake case

File: nxc/protocols/smb/kerberos.py
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:196]: Needs snake case

File: nxc/protocols/smb/kerberos.py
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:208]: Needs snake case

File: nxc/protocols/smb/kerberos.py
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:210]: Needs snake case

File: nxc/protocols/smb/kerberos.py
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:220]: Needs snake case

File: nxc/protocols/smb/kerberos.py
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:227]: Needs snake case

File: nxc/protocols/smb/kerberos.py
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:241]: Needs snake case

File: nxc/protocols/smb/kerberos.py
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:18]: useCache needs snakecase

File: nxc/protocols/smb/kerberos.py
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:20]: Should just be "tgt"

File: nxc/protocols/smb/kerberos.py
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:19]: Shouldn't have to do this, you can just do `from nxc.logger import nxc_logger` and then use that.

File: nxc/protocols/smb/kerberos.py
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:89]: Needs snake case

File: nxc/protocols/smb/kerberos.py
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:91]: Needs snake case

File: nxc/protocols/smb/kerberos.py
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:98]: Needs snake case

File: nxc/protocols/smb/kerberos.py
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:100]: Needs snake case

File: nxc/protocols/smb/kerberos.py
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:108]: Needs snake case

File: nxc/protocols/smb/kerberos.py
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:114]: Needs snake case

File: nxc/protocols/smb/kerberos.py
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:123]: Needs snake case

File: nxc/protocols/smb/kerberos.py
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:263]: Needs snake case

File: nxc/protocols/smb/kerberos.py
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:268]: Needs snake case

File: nxc/protocols/smb/kerberos.py
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:270]: Needs snake case

File: nxc/protocols/smb/kerberos.py
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:272]: Needs snake case

File: nxc/protocols/smb/kerberos.py
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:152]: Needs snake case

File: nxc/protocols/smb/kerberos.py
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:150]: Needs snake case

File: nxc/protocols/smb/kerberos.py
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:148]: Needs snake case

File: nxc/protocols/smb/kerberos.py
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:143]: Needs snake case

File: nxc/protocols/smb/kerberos.py
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:62]: Needs snake case

File: nxc/protocols/smb/kerberos.py
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:74]: Needs snake case

File: nxc/protocols/smb/kerberos.py
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:68]: Needs snake case

File: nxc/protocols/smb/kerberos.py
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:76]: Needs snake case

File: nxc/protocols/smb/kerberos.py
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:41]: Needs snake case

File: nxc/protocols/smb/kerberos.py
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:36]: Needs snake case

File: nxc/protocols/smb/kerberos.py
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:35]: should be snake case, but "tgt" is taken above. Maybe rename the one above and set this to "tgt"

File: nxc/protocols/smb/kerberos.py
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:25]: Needs snake case

**Review by Marshall-Hallenbeck — CHANGES_REQUESTED** [review]
Mostly a bunch of variable naming to conform to proper Pythonic naming, but also pointed out importing the nxc logger.

**Comment by zblurx** [comment]
All good now I think

**Comment by NeffIsBack** [comment]
There are still a few single quotes but we should probably wait until #35 is merged, a lot of it is fixed in there

**Comment by Marshall-Hallenbeck** [comment]
@zblurx #35 got merged in, if you can take care of the few conflicts here it should be good to merge.

**Comment by NeffIsBack** [comment]
There must be an `hasattribute` check on the delegation arg. Otherwise we get that error on other protocols:
image

**Comment by NeffIsBack** [comment]
image
Working like a charm, except `--ntds`. @zblurx is this fixable?
image

**Comment by zblurx** [comment]
So for the `--ntds`, the problem is with impacket secretsdump. Actually secretsdump needs a cifs ticket AND an ldap ticket, but the `--delegate` will only get a cifs ticket. So my workaround is :
- 1: You use `--delegate` to do a `--lsa` and get the nthash of the DC computer account
- 2: You use the DC computer account to do a dcsync with `--ntds`

**Comment by NeffIsBack** [comment]
Working flawlessly
image

**Review by NeffIsBack — APPROVED** [review]
LGTM 🎉

**Review by mpgn — APPROVED** [review]

**Review by Marshall-Hallenbeck — APPROVED** [review]
LGTM

**Pull request merged**
